### PR TITLE
Omptimized SourceByEmitter builders

### PR DESCRIPTION
**Pull request by arkivanov**
Our Scrabble results are already reduced from ~205 ms to ~131 ms on my machine. This PR improves it further to ~124 ms.

image

### CODE REVIEW

**Comment by CherryPerry** [comment]
Check diff.
+1.11 KB buildSrc/buildSrc/build/classes/kotlin/main/Deps$Android$Support.class

**Comment by arkivanov** [comment]
Oops, will fix it this evening

**Comment by arkivanov** [comment]
Fixed

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt [hunk 1]
```diff
@@ -2,41 +2,32 @@ package com.badoo.reaktive.single
...
-                            observer.onError(error)
-                        } finally {
-                            disposableWrapper.dispose()
-                        }
+                    doIfNotDisposed(dispose = true) {
+                        observer.onError(error)
                     }
                 }
 
+                @Suppress("RedundantOverride") // IDEA complains that setDisposable is not implemented
```
Comment (CherryPerry) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt]: Only here and not in completable, maybe and observable?

Comment (arkivanov) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt]: Mistake, will fix

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/CompletableByEmitter.kt [hunk 2]
```diff
@@ -2,41 +2,29 @@ package com.badoo.reaktive.completable
...
 fun completable(onSubscribe: (emitter: CompletableEmitter) -> Unit): Completable =
     completableUnsafe { observer ->
-        val disposableWrapper = DisposableWrapper()
-        observer.onSubscribe(disposableWrapper)
-
         val emitter =
-            object : CompletableEmitter {
-                override val isDisposed: Boolean get() = disposableWrapper.isDisposed
-
+            object : DisposableWrapper(), CompletableEmitter {
```
Comment (CherryPerry) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/CompletableByEmitter.kt]: What if we introduce new class?
```kotlin
abstract class ObserverDisposableWrapper: DisposableWrapper(), InterfaceThatHasSetDisposableFunction {
    override fun setDisposable(disposable: Disposable) {
        set(disposable)
    }
}
```

Comment (arkivanov) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/CompletableByEmitter.kt]: Makes sense, will do for all

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt [hunk 3]
```diff
@@ -1,35 +1,33 @@
...
+        val emitter = DisposableEmitter(observer)
         observer.onSubscribe(emitter)
+        emitter.tryCatch(block = { onSubscribe(emitter) })
+    }
 
-        try {
-            onSubscribe(emitter)
-        } catch (e: Throwable) {
-            emitter.onError(e)
+private class DisposableEmitter<T>(
```
Comment (CherryPerry) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt]: No, what is the point to do it? I wanted to remove copy-pasted
```kotlin
override fun setDisposable(disposable: Disposable) {
        set(disposable)
}
```
block by delegating it to common base class.

Comment (arkivanov) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt]: I don't understand, please elaborate

Comment (CherryPerry) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt]: Check original comment. I proposed to introduce common base class for all 4 cases:
```kotlin
abstract class EmitterDisposableWrapper: DisposableWrapper(), Emitter {
    override fun setDisposable(disposable: Disposable) {
        set(disposable)
    }
}
...
val emitter = object : EmitterDisposableWrapper(), CompletableEmitter { }
val emitter = object : EmitterDisposableWrapper(), MaybeEmitter { }
...
```

Comment (arkivanov) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt]: Fixed! Thanks!

**Pull request merged**